src/main/java/namor/template/calculator/RPNcalculator.java

```diff
@@ -1,66 +1,117 @@
+        if (stack.equals(ArrayStack.class)){
+            numbers = new ArrayStack<>();
+        } else if (stack.equals(MapBasedStack.class)){
+            numbers = new MapBasedStack<>();
+        }
+        this.type = type;
+    }
 
-    public int calculate(String arithmeticExpression) {
-        ArrayStack numbers = new ArrayStack();
+    private Class<T> getMyType() {
+        return this.type;
+    }
+
+    public T calculate(String arithmeticExpression) {
 
         for (String operationElement : arithmeticExpression.split(" ")) {
             if (Sign.isSign(operationElement)) {
                 doArithmeticOperation(operationElement, numbers);
             } else {
-                numbers.push(Integer.parseInt(operationElement));
+                convertStringToNumber(getMyType(), numbers, operationElement);
             }
         }
         return numbers.pop();
     }
 
-    private void doArithmeticOperation(String arithmeticOperator, ArrayStack numbers) {
+    private void doArithmeticOperation(String arithmeticOperator, Stack<T> numbers) {
         switch (Sign.getValue(arithmeticOperator)) {
             case ADDITION:
-                numbers.push(addition(numbers));
+                numbers.push(addition(numbers.pop(), numbers.pop()));
                 break;
             case SUBTRACTION:
-                numbers.push(subtraction(numbers));
+                numbers.push(subtraction(numbers.pop(), numbers.pop()));
                 break;
             case MULTIPLICATION:
-                numbers.push(multiplication(numbers));
+                numbers.push(multiplication(numbers.pop(), numbers.pop()));
                 break;
             case DIVISION:
-                numbers.push(division(numbers));
+                numbers.push(division(numbers.pop(), numbers.pop()));
                 break;
         }
     }
 
-    private final int addition(ArrayStack numbers) {
-        final int secondNumber = numbers.pop();
-        final int firstNumber = numbers.pop();
-
-        return firstNumber + secondNumber;
+    private void convertStringToNumber(Class<T> type, Stack<T> numbers, String operationElement) {
+        if(type == Integer.class) {
+            numbers.push((T)(Integer)(Integer.parseInt(operationElement)));
+        } else if(type == Double.class) {
+            numbers.push((T)(Double)(Double.parseDouble(operationElement)));
+        } else if(type == Float.class) {
+            numbers.push((T)(Float)(Float.parseFloat(operationElement)));
+        } else {
+            numbers.push((T)(Long)(Long.parseLong(operationElement)));
+        }
     }
 
-    private final int subtraction(ArrayStack numbers) {
-        final int secondNumber = numbers.pop();
-        final int firstNumber = numbers.pop();
-
-        return firstNumber - secondNumber;
+    private final <T> T addition(T firstNumber, T secondNumber) {
+        if (firstNumber.getClass() == Integer.class) {
```

Comment:
All this if-else to check the type are useless. There shouldn't be logic like that. You don't understand how primitives works. Shortly in all this operations should not be statements and only one line like `return first + second`

Reply:
I don't understand how i can change this method. Can you prompt me what information I should look for?